### PR TITLE
Decadac in Custom-Instr

### PR DESCRIPTION
**Pull request by ThorvaldLarsen**
Making the the decadac driver more streamlined so we only need to change things in the config file.

The code is not pretty with 'exec' but dont know a better way to code it - Just want to have all settings in the config file rather than both places.

Note new format in the instr.config file.

### CODE REVIEW

File: qdev_wrappers/T3/customised_instruments.py [hunk 1]
```diff
@@ -227,78 +227,66 @@ def __init__(self, name, address, config, **kwargs):
...
+            self.channels[chan].volt.set_step(step)
+            self.channels[chan].volt.set_delay(delay)
+            
+            
+            rangemin = float(config_settings[6])
+            rangemax = float(config_settings[7])
             vldtr = vals.Numbers(rangemin, rangemax)
             self.channels[chan].volt.set_validator(vldtr)
+            if config_settings[8] == 'fine':
+                if np.mod(chan,4)<3:
```
Comment (Dominik-Vogel) [inline qdev_wrappers/T3/customised_instruments.py]: Why do you exclude the last channel of every slot?

Comment (ThorvaldLarsen) [reply qdev_wrappers/T3/customised_instruments.py]: that's a mistake. Should have been less than or equal to 3 (or <4). Haven't noticed as we have never used the last channel.

Comment (Dominik-Vogel) [reply qdev_wrappers/T3/customised_instruments.py]: But isn't mod(x,3)∈{0,1,2,3}? So that the if statement is always true?

Comment (ThorvaldLarsen) [reply qdev_wrappers/T3/customised_instruments.py]: I take back what I said before. I think my idea was to not define the last two channels when in 'fine' mode but then made a wrong implementation. Also this isn't really needed as fine_volt already throws an error for the last two channels of each slot.
Maybe just remove the if statement for both 'coarse' and 'fine' mode? It definitely seems unnecessary.

Comment (Dominik-Vogel) [reply qdev_wrappers/T3/customised_instruments.py]: ok

File: qdev_wrappers/T3/customised_instruments.py [hunk 2]
```diff
@@ -227,78 +227,66 @@ def __init__(self, name, address, config, **kwargs):
...
-                raise ValueError(
-                    "Expected: min max. Got {}".format(chan_range))
-            else:
-                rangemin = float(range_minmax[0])
-                rangemax = float(range_minmax[1])
+            config_settings = config_file[str(chan)].split(",")
+            
+            step = float(config_settings[4])
+            delay = float(config_settings[5])
+            self.channels[chan].volt.set_step(step)
```
Comment (Dominik-Vogel) [inline qdev_wrappers/T3/customised_instruments.py]: Why do you only set the step for the volt and not for the fine_volt?

Comment (ThorvaldLarsen) [reply qdev_wrappers/T3/customised_instruments.py]: my thinking was that fine_volt refers back to volt. So by setting volt on all channels I will also define step on the fine_volt implicitly.

File: qdev_wrappers/T3/customised_instruments.py [hunk 3]
```diff
@@ -243,49 +243,40 @@ def __init__(self, name, address, config, **kwargs):
...
-                        exec('self.{} = self.channels[chan].volt'.format(config_settings[0]))
-                        exec('self.{}.label = \'{}\''.format(config_settings[0],config_settings[1]))
-                        exec('self.{}.unit = \'{}\''.format(config_settings[0],config_settings[2]))
-
+            fine_mode = config_settings[8]
+
+            if  fine_mode == 'fine':
+                param = channel.fine_volt
+            elif fine_mode == 'coarse':
+                param = channel.fine_volt
```
Comment (ThorvaldLarsen) [inline qdev_wrappers/T3/customised_instruments.py]: think this should be channel.volt for coarse mode.

File: qdev_wrappers/T3/customised_instruments.py [hunk 4]
```diff
@@ -243,49 +243,40 @@ def __init__(self, name, address, config, **kwargs):
...
+            else:
+                pass #  raise exception
+
+            param.label = label
+            param.unit = unit
+            param.set_step(step)
+            param.set_delay(delay)
```
Comment (ThorvaldLarsen) [inline qdev_wrappers/T3/customised_instruments.py]: set_step and set_delay should refer to channel.volt both in fine and coarse mode.

File: qdev_wrappers/T3/customised_instruments.py (hunk 3 above)
Comment (Dominik-Vogel) [reply qdev_wrappers/T3/customised_instruments.py]: well spotted.

File: qdev_wrappers/T3/customised_instruments.py (hunk 4 above)
Comment (Dominik-Vogel) [reply qdev_wrappers/T3/customised_instruments.py]: done

**Comment by Dominik-Vogel** [comment]
Did you change any security settings @ThorvaldLarsen? I cannot push the changes....

**Comment by ThorvaldLarsen** [comment]
Haven't changed anything... It is saying that maintainers are allowed to commit changes.

File: qdev_wrappers/T3/customised_instruments.py [hunk 5]
```diff
@@ -322,8 +241,7 @@ def _set_fine_voltage(self, voltage):
             fine_chan = 3
         else:
             raise RuntimeError("Fine mode only works for Chan 0 and 1")
-        coarse_part = self._dac_code_to_v(
-            self._dac_v_to_code(round(voltage,2)-0.01) )
+        coarse_part = self._dac_code_to_v(self._dac_v_to_code(voltage-0.001))
```
Comment (Dominik-Vogel) [inline qdev_wrappers/T3/customised_instruments.py]: @ThorvaldLarsen I think this should not be changed... I don't know how this got in here. But as I remember it, does the originalversion with the rounding come from T3. Can you confirm this?

File: qdev_wrappers/T3/customised_instruments.py [hunk 6]
```diff
@@ -332,31 +250,78 @@ def _set_fine_voltage(self, voltage):
         self.volt.set(coarse_part)
         slot.channels[fine_chan].volt.set(fine_scaled)
 
-    def set_all(self, voltage_value, set_dcbias=False):
```
Comment (Dominik-Vogel) [inline qdev_wrappers/T3/customised_instruments.py]: @ThorvaldLarsen  do we want to remove this?

Comment (ThorvaldLarsen) [reply qdev_wrappers/T3/customised_instruments.py]: It is some old stuff that we never used. Lets remove it for now and get a clean driver to start from. Then maybe later there might be a few functions we want to add again.

File: qdev_wrappers/T3/customised_instruments.py (hunk 5 above)
Comment (ThorvaldLarsen) [reply qdev_wrappers/T3/customised_instruments.py]: I think this change is good. It avoids large voltage jumps in between setting the coarse and fine parts of the voltage. Not sure how it got here but reading it I like the change.

**Pull request merged**